default/AI/AIstate.py

```diff
@@ -268,6 +273,39 @@ def assess_planet_threat(self, pid, sightingAge=0):
         infra = 0  # doesn't really contribute to combat since damage to shields & 'defense' done first
         return {'overall': defense*(defense + shields + infra), 'attack': defense, 'health': (defense + shields + infra)}
 
+    def log_immobile_fleet(self, fleet, system, rating):
+        if not fleet or not system:
+            return
+        universe = fo.getUniverse()
+        print "Immobile fleet %s at %s has rating %.1f and %d ships:" % (fleet, system, rating.get('overall', 0), len(fleet.shipIDs))
+        for ship_id in fleet.shipIDs:
+            this_ship = universe.getShip(ship_id)
+            this_design = fo.getShipDesign(this_ship.designID)
+            print "\t\t -- %s %d of design %s" % (["ship", "monster"][this_ship.isMonster], ship_id, this_design.name(False))
+
+    def assess_enemy_supply(self):
+        """
+        Assesses where enemy empires have Supply
+        :return:a tuple of 2 dicts, each of which is keyed by system id, and each of which is a list of empire ids
+        1st dict-- enemies that actually have supply at this system
+        2nd dict-- enemies that have supply within 2 jumps from this system (if they clear obstructions)
+        """
+        enemy_ids = [_id for _id in fo.allEmpireIDs() if _id != fo.empireID()]
+        actual_supply = {}
+        near_supply = {}
+        for enemy_id in enemy_ids:
+            this_enemy = fo.getEmpire(enemy_id)
+            if not this_enemy:
+                print "Could not retrieve empire for empire id %d" % enemy_id  # do not spam chat_error with this
```

Comment:
In an observer or moderator multiplayer game, the observer/moderator is assigned empire ID  -1, and since the observer/moderator is one of the entries in map of player_info, this gets included in the allEmpireIDs results.  So I am not inclined to write it to stderr.  But just in case there is a real problem sometime I am inclined to log it as it is now instead of just ignoring it.  I suppose the logging could be only if the empireID were not -1, and in that case sending it to the error log would probably be appropriate as well.  And/or you chould check if there really was a moderator/observer in the game.  But doing all that seems a bit of overkill.